dowhy/causal_refuters/add_unobserved_common_cause.py

```diff
@@ -119,634 +123,821 @@ def __init__(self, *args, **kwargs):
+    :plugin_reisz: bool: PENDING_PARAM_DOC
+    """
 
-    def include_confounders_effect(self, new_data, kappa_t, kappa_y):
-        """
-        This function deals with the change in the value of the data due to the effect of the unobserved confounder.
-        In the case of a binary flip, we flip only if the random number is greater than the threshold set.
-        In the case of a linear effect, we use the variable as the linear regression constant.
+    # If the estimator used is LinearDML, partially linear sensitivity analysis will be automatically chosen
+    if isinstance(estimate.estimator, dowhy.causal_estimators.econml.Econml):
+        if estimate.estimator._econml_methodname == "econml.dml.LinearDML":
+            analyzer = PartialLinearSensitivityAnalyzer(
+                estimator=estimate._estimator_object,
+                observed_common_causes=estimate.estimator._observed_common_causes,
+                treatment=estimate.estimator._treatment,
+                outcome=estimate.estimator._outcome,
+                alpha_s_estimator_param_list=alpha_s_estimator_param_list,
+                g_s_estimator_list=g_s_estimator_list,
+                g_s_estimator_param_list=g_s_estimator_param_list,
+                effect_strength_treatment=kappa_t,
+                effect_strength_outcome=kappa_y,
+                benchmark_common_causes=benchmark_common_causes,
+                frac_strength_treatment=frac_strength_treatment,
+                frac_strength_outcome=frac_strength_outcome,
+            )
+            analyzer.check_sensitivity(plot=plot_estimate)
+            return analyzer
 
-        :param new_data: pandas.DataFrame: The data to be changed due to the effects of the unobserved confounder.
-        :param kappa_t: numpy.float64: The value of the threshold for binary_flip or the value of the regression coefficient for linear effect.
-        :param kappa_y: numpy.float64: The value of the threshold for binary_flip or the value of the regression coefficient for linear effect.
+    analyzer = NonParametricSensitivityAnalyzer(
+        estimator=estimate.estimator,
+        observed_common_causes=estimate.estimator._observed_common_causes,
+        treatment=estimate.estimator._treatment,
+        outcome=estimate.estimator._outcome,
+        alpha_s_estimator_list=alpha_s_estimator_list,
+        alpha_s_estimator_param_list=alpha_s_estimator_param_list,
+        g_s_estimator_list=g_s_estimator_list,
+        g_s_estimator_param_list=g_s_estimator_param_list,
+        effect_strength_treatment=kappa_t,
+        effect_strength_outcome=kappa_y,
+        benchmark_common_causes=benchmark_common_causes,
+        frac_strength_treatment=frac_strength_treatment,
+        frac_strength_outcome=frac_strength_outcome,
+        theta_s=estimate.value,
+        plugin_reisz=plugin_reisz,
+    )
+    analyzer.check_sensitivity(plot=plot_estimate)
+    return analyzer
+
+
+def refute_add_unobserved_common_cause(
```

Comment:
with the functional API, is there a way to ensure that any future refuter functions follow the same basic signature?
E.g., all refute functions should have a data, target_estimand and estimate? I'm thinking of making it easier to for new contributors to know the basic protocol that they need to adhere to. 
something like `def refute_*(data, target_estimand, estimate, *args, **kwargs)`

Reply:
I believe there is not a specific way of doing this, however the main `refute_estimate(data, target_estimand, estimate, ...)` partially serves this purpose, in the future if any developer wants the refuter to be callable through this function then the signature must comply with the type: `Callable[..., Union[CausalRefutation, List[CausalRefutation]]]`, which at least hints the developer to return a CausalRefutation or list of CausalRefutation as result. Or we can create as you say a protocol like this:

```python
class Refuter(Protocol):
    def __call__(self, data: pd.DataFrame, target_estimand: IdentifiedEstimand, estimate: CausalEstimate, *args, **kwargs) -> Union[CausalRefutation, List[CausalRefutation]]
```

and then use it as the type hint in the `refute_estimate(..., refuters=List[Refuter])`

https://mypy.readthedocs.io/en/latest/protocols.html#callback-protocols


I don't really think this is super useful as we also have PR mechanism that we could use to enforce compatibility, but if you consider we should add it like this, then let me know and I'll include it in this PR.

Reply:
Got, thanks for clarifying. I do agree that adding a class as protocol may not be needed. We can do it through PR reviews and docs to communicate the expected signature.

Perhaps the flexibility will also be useful for any future method that may not require to take in the usual parameters.